File: developer-docs-site/static/examples/python/first_coin.py
#!/usr/bin/env python3

# Copyright (c) Aptos
# SPDX-License-Identifier: Apache-2.0

from hello_blockchain import HelloBlockchainClient
from typing import Optional
import sys

from first_transaction import Account, FaucetClient, RestClient, TESTNET_URL, FAUCET_URL

class FirstCoin(RestClient):

#:!:>section_1
    def initialize_coin(self, account_from: Account) -> Optional[str]:
        """ Initialize a new coin with the given coin type. """
        payload = {
            "type": "script_function_payload",
            "function": "0x1::ManagedCoin::initialize",
            "type_arguments": [f"0x{account_from.address()}::MoonCoin::MoonCoin"],
            "arguments": [
                "Moon Coin".encode("utf-8").hex(),
                "MOON".encode("utf-8").hex(),
                "6",
                False
            ]
        }
        res = self.execute_transaction_with_payload(account_from, payload)
        return str(res["hash"])
#<:!:section_1

#:!:>section_2
    def register_coin(self, account_receiver: Account, coin_type_address: str) -> str:
        """ Register the receiver account to receive transfers for the new coin. """

        payload = {
            "type": "script_function_payload",
            "function": "0x1::Coin::register",
            "type_arguments": [f"0x{coin_type_address}::MoonCoin::MoonCoin"],
            "arguments": []
        }
        res = self.execute_transaction_with_payload(account_receiver, payload)
        return str(res["hash"])
#<:!:section_2

#:!:>section_3
    def mint_coin(
        self,
        account_coin_owner: Account,
        receiver_address: str,
        amount: int
    ) -> str:
        """ Register the receiver account to receive transfers for the new coin. """

        payload = {
            "type": "script_function_payload",
            "function": "0x1::ManagedCoin::mint",
            "type_arguments": [f"0x{account_coin_owner.address()}::MoonCoin::MoonCoin"],
            "arguments": [
                receiver_address,
                f"{amount}"
            ]
        }
        res = self.execute_transaction_with_payload(account_coin_owner, payload)
        return str(res["hash"])
#<:!:section_3

#:!:>section_4
    def get_balance(
        self,
        account_address: str,
        coin_type_address: str,
    ) -> str:
        """ Returns the coin balance of the given account """

        return self.account_resource(account_address, f"0x1::Coin::CoinStore<0x{coin_type_address}::MoonCoin::MoonCoin>")
#<:!:section_4

if __name__ == "__main__":
    assert len(sys.argv) == 2, "Expecting an argument that points to the helloblockchain module"

    client = FirstCoin(TESTNET_URL)
    faucet_client = FaucetClient(FAUCET_URL, client)

    alice = Account()
    bob = Account()

    print("\n=== Addresses ===")
    print(f"Alice: {alice.address()}")
    print(f"Bob: {bob.address()}")

    faucet_client.fund_account(alice.address(), 10_000_000)
    faucet_client.fund_account(bob.address(), 10_000_000)

    input("\nUpdate the module with Alice's address, build, copy to the provided path, and press enter.")
    module_path = sys.argv[1]
    with open(module_path, "rb") as f:
        module_hex = f.read().hex()

    print("Publishing MoonCoinType module...")
    hello_blockchain_client = HelloBlockchainClient(TESTNET_URL)
    tx_hash = hello_blockchain_client.publish_module(alice, module_hex)
    hello_blockchain_client.wait_for_transaction(tx_hash)

    print("Alice will initialize the new coin")
    tx_hash = client.initialize_coin(alice)
    client.wait_for_transaction(tx_hash)
    
    print("Bob registers the newly created coin so he can receive it from Alice")
    tx_hash = client.register_coin(bob, alice.address())
    client.wait_for_transaction(tx_hash)
    print(f"Bob's initial MoonCoinType balance: {client.get_balance(bob.address(), alice.address())}")
    
    print("Alice mints Bob some of the new coin")
    tx_hash = client.mint_coin(alice, bob.address(), 100)
    client.wait_for_transaction(tx_hash)
    print(f"Bob's updated MoonCoinType balance: {client.get_balance(bob.address(), alice.address())}")

File: developer-docs-site/static/examples/typescript/first_coin.ts
// Copyright (c) The Aptos Foundation
// SPDX-License-Identifier: Apache-2.0

import assert from "assert";
import fs from "fs";
import { NODE_URL, FAUCET_URL } from "./first_transaction";
import { AptosAccount, AptosClient, TxnBuilderTypes, BCS, MaybeHexString, HexString, FaucetClient } from "aptos";
import { publishModule } from "./hello_blockchain";

const readline = require("readline").createInterface({
  input: process.stdin,
  output: process.stdout,
});

//:!:>section_1
const client = new AptosClient(NODE_URL);
/** Initializes the new coin */
async function initializeCoin(accountFrom: AptosAccount, coinTypeAddress: HexString): Promise<string> {
  const token = new TxnBuilderTypes.TypeTagStruct(
    TxnBuilderTypes.StructTag.fromString(`${coinTypeAddress.hex()}::moon_coin::MoonCoin`),
  );

  const serializer = new BCS.Serializer();
  serializer.serializeBool(false);

  const entryFunctionPayload = new TxnBuilderTypes.TransactionPayloadEntryFunction(
    TxnBuilderTypes.EntryFunction.natural(
      "0x1::managed_coin",
      "initialize",
      [token],
      [BCS.bcsSerializeStr("Moon Coin"), BCS.bcsSerializeStr("MOON"), BCS.bcsSerializeUint64(6), serializer.getBytes()],
    ),
  );

  const [{ sequence_number: sequenceNumber }, chainId] = await Promise.all([
    client.getAccount(accountFrom.address()),
    client.getChainId(),
  ]);

  const rawTxn = new TxnBuilderTypes.RawTransaction(
    TxnBuilderTypes.AccountAddress.fromHex(accountFrom.address()),
    BigInt(sequenceNumber),
    entryFunctionPayload,
    1000n,
    1n,
    BigInt(Math.floor(Date.now() / 1000) + 10),
    new TxnBuilderTypes.ChainId(chainId),
  );

  const bcsTxn = AptosClient.generateBCSTransaction(accountFrom, rawTxn);
  const pendingTxn = await client.submitSignedBCSTransaction(bcsTxn);

  return pendingTxn.hash;
}
//<:!:section_1

//:!:>section_2
/** Receiver needs to register the coin before they can receive it */
async function registerCoin(coinReceiver: AptosAccount, coinTypeAddress: HexString): Promise<string> {
  const token = new TxnBuilderTypes.TypeTagStruct(
    TxnBuilderTypes.StructTag.fromString(`${coinTypeAddress.hex()}::moon_coin::MoonCoin`),
  );

  const entryFunctionPayload = new TxnBuilderTypes.TransactionPayloadEntryFunction(
    TxnBuilderTypes.EntryFunction.natural("0x1::coins", "register", [token], []),
  );

  const [{ sequence_number: sequenceNumber }, chainId] = await Promise.all([
    client.getAccount(coinReceiver.address()),
    client.getChainId(),
  ]);

  const rawTxn = new TxnBuilderTypes.RawTransaction(
    TxnBuilderTypes.AccountAddress.fromHex(coinReceiver.address()),
    BigInt(sequenceNumber),
    entryFunctionPayload,
    1000n,
    1n,
    BigInt(Math.floor(Date.now() / 1000) + 10),
    new TxnBuilderTypes.ChainId(chainId),
  );

  const bcsTxn = AptosClient.generateBCSTransaction(coinReceiver, rawTxn);
  const pendingTxn = await client.submitSignedBCSTransaction(bcsTxn);

  return pendingTxn.hash;
}
//<:!:section_2

//:!:>section_3
/** Mints the newly created coin to a specified receiver address */
async function mintCoin(
  coinOwner: AptosAccount,
  coinTypeAddress: HexString,
  receiverAddress: HexString,
  amount: number,
): Promise<string> {
  const token = new TxnBuilderTypes.TypeTagStruct(
    TxnBuilderTypes.StructTag.fromString(`${coinTypeAddress.hex()}::moon_coin::MoonCoin`),
  );

  const entryFunctionPayload = new TxnBuilderTypes.TransactionPayloadEntryFunction(
    TxnBuilderTypes.EntryFunction.natural(
      "0x1::managed_coin",
      "mint",
      [token],
      [BCS.bcsToBytes(TxnBuilderTypes.AccountAddress.fromHex(receiverAddress.hex())), BCS.bcsSerializeUint64(amount)],
    ),
  );

  const [{ sequence_number: sequenceNumber }, chainId] = await Promise.all([
    client.getAccount(coinOwner.address()),
    client.getChainId(),
  ]);

  const rawTxn = new TxnBuilderTypes.RawTransaction(
    TxnBuilderTypes.AccountAddress.fromHex(coinOwner.address()),
    BigInt(sequenceNumber),
    entryFunctionPayload,
    1000n,
    1n,
    BigInt(Math.floor(Date.now() / 1000) + 10),
    new TxnBuilderTypes.ChainId(chainId),
  );

  const bcsTxn = AptosClient.generateBCSTransaction(coinOwner, rawTxn);
  const pendingTxn = await client.submitSignedBCSTransaction(bcsTxn);
  return pendingTxn.hash;
}
//<:!:section_3

//:!:>section_4
/** Return the balance of the newly created coin */
async function getBalance(accountAddress: MaybeHexString, coinTypeAddress: HexString): Promise<string | number> {
  try {
    const resource = await client.getAccountResource(
      accountAddress,
      `0x1::coin::CoinStore<${coinTypeAddress.hex()}::moon_coin::MoonCoin>`,
    );

    return parseInt((resource.data as any)["coin"]["value"]);
  } catch (_) {
    return 0;
  }
}
//<:!:section_4

/** run our demo! */
async function main() {
  assert(process.argv.length == 3, "Expecting an argument that points to the moon coin module");

  const client = new AptosClient(NODE_URL);
  const faucetClient = new FaucetClient(NODE_URL, FAUCET_URL);

  // Create two accounts, Alice and Bob, and fund Alice but not Bob
  const alice = new AptosAccount();
  const bob = new AptosAccount();

  console.log("\n=== Addresses ===");
  console.log(`Alice: ${alice.address()}`);
  console.log(`Bob: ${bob.address()}`);

  await faucetClient.fundAccount(alice.address(), 5_000);
  await faucetClient.fundAccount(bob.address(), 5_000);

  await new Promise<void>((resolve) => {
    readline.question(
      "Update the CoinType module with Alice's address, build, copy to the provided path, and press enter.",
      () => {
        resolve();
        readline.close();
      },
    );
  });
  const modulePath = process.argv[2];
  const moduleHex = fs.readFileSync(modulePath).toString("hex");

  console.log("Publishing MoonCoinType module...");
  let txHash = await publishModule(alice, moduleHex);
  await client.waitForTransaction(txHash);

  console.log("Alice will initialize the new coin");
  txHash = await initializeCoin(alice, alice.address());
  await client.waitForTransaction(txHash);

  console.log("Bob registers the newly created coin so he can receive it from Alice");
  txHash = await registerCoin(bob, alice.address());
  await client.waitForTransaction(txHash);
  console.log(`Bob's initial balance: ${await getBalance(bob.address(), alice.address())}`);

  console.log("Alice mints Bob some of the new coin");
  txHash = await mintCoin(alice, alice.address(), bob.address(), 100);
  await client.waitForTransaction(txHash);
  console.log(`Bob's new balance: ${await getBalance(bob.address(), alice.address())}`);
}

if (require.main === module) {
  main().then((resp) => console.log(resp));
}
